### ISSUE TITLE
plugin/forward: transport.Yield may block

### ISSUE THREAD

**Issue by miekg** [coredns/coredns#3335]
Happen to look more in forward.

The yielding of connection blocks in *forward* until the connection is ready to be returned. This can block for a long time until the connection spike is over, because `connManager` is busy looking for connection and doing other stuff.

Changing to something like this seem prudent:
~~~
// Yield return the connection to transport for reuse.
func (t *Transport) Yield(c *dns.Conn) {
	// Make ths non-blocking, because in the case of a very busy forwarder we will *block* on this yield. This
	// blocks the outer go-routine and stuff will just pile up.
	select {
	case t.yield <- c:
		return
	case <-time.After(10 * time.Millisecond):
		log.Warning("failed to yield connection")
	}
}
~~~

*forward* should also just use `*persistConn` in more places - I think that would simplify the code.

**Comment by rdrozhdzh** [comment]
alternatively/additionally you can consider making `Transport.yield` channel buffered.

One more solution to make `connManager()` faster is #1941

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/forwar..." ]
>alternatively/additionally you can consider making `Transport.yield` channel buffered.

No, this is dangerous because it means the channel fills up at the worst 
possible moment and the test probably don't have that scale, so they don't catch 
it.

>One more solution to make `connManager()` faster is #1941

because the removal of the string comparison? How much faster does that get for 
a localhost <-> localhost proxy setup?

**Comment by rdrozhdzh** [comment]
> because the removal of the string comparison? 

Not only string comparison. AFAIK map first calculates hash of string to determine a bucket inside the map and then it looks through the values in the bucket and compares the keys because different keys can result into the same hashes.

> How much faster does that get for a localhost <-> localhost proxy setup?

This still needs some more effort to write a benchmark. Probably I could find time in near future and continue work on #1941 if you are interested in this change

**Comment by miekg** [comment]
I think just moving a const value for `udp`, `tcp` and `tcp-tls` and removing that map would be a much smaller change with the same benefits.

Alternative we just write it out 3 times: https://github.com/coredns/coredns/pull/3334/files#diff-c641d6db9fa84aad4501394efccea2dbR22

**Comment by miekg** [comment]
I don't have any proof; but this might explain some of the things seen #2593 - i.e in the following scenario:

1. dial being the first case in the select in `transport.Dial` it will always be done first (if I remember the golang spec correctly)
2. traffic spike: get a whole bunch of sockets, do your thing
3. spike continues; dial still busy (See 1.)
4. yield hangs on the yield channel send
5. go-routines pile up
6. mayhem ensues and memory, killing a couple of coredns pods -> cascading failure

**Comment by rdrozhdzh** [comment]
> it will always be done first (if I remember the golang spec correctly)

Not sure about this. From https://golang.org/ref/spec#Select_statements

Execution of a "select" statement proceeds in several steps:
1. For all the cases in the statement, the channel operands of receive operations and the channel and right-hand-side expressions of send **statements are evaluated exactly once, in source order, upon entering the "select" statement. The result is a set of channels to receive from or send to, and the corresponding values to send**. Any side effects in that evaluation will occur irrespective of which (if any) communication operation is selected to proceed. Expressions on the left-hand side of a RecvStmt with a short variable declaration or assignment are not yet evaluated.
2.    If one or more of the communications can proceed, **a single one that can proceed is chosen via a uniform pseudo-random selection**. Otherwise, if there is a default case, that case is chosen. If there is no default case, the "select" statement blocks until at least one of the communications can proceed.
3.    Unless the selected case is the default case, the respective communication operation is executed.
4.    If the selected case is a RecvStmt with a short variable declaration or an assignment, the left-hand side expressions are evaluated and the received value (or values) are assigned.
5.    The statement list of the selected case is executed.

**Comment by rdrozhdzh** [comment]
Regarding #2593 I suspect that the problem is that coredns receives new queries faster than upstreams can process (resolve) them. Therefore a lot of goroutines get hung waiting (until timeout) for response from upstreams.

**Comment by miekg** [comment]
ack, so it's random. Note that with some (now removed) logging, there was def. blocking going on there while the server it busy. Slow upstream should be testable with the erratic plugin and delay

**Comment by miekg** [comment]
fixed via #3336
